tests/server.c

```diff
@@ -0,0 +1,359 @@
+  struct timespec ts;
+  enum exchange_state state = READING_REQUEST;
+
+  fprintf(stderr, "accepted conn on fd %d\n", conn->fd);
+
+  for(;;) {
+    FD_ZERO(&read_fds);
+    if (rustls_connection_wants_read(rconn)) {
+      FD_SET(sockfd, &read_fds);
+    }
+    FD_ZERO(&write_fds);
+    if (rustls_connection_wants_write(rconn)) {
+      FD_SET(sockfd, &write_fds);
+    }
+
+    result = select(sockfd + 1, &read_fds, &write_fds, NULL, NULL);
+    if(result == -1) {
+      perror("select");
+      goto cleanup;
+    }
+    if(result == 0) {
+      fprintf(stderr, "no fds from select, sleeping\n");
+      ts.tv_sec = 0;
+      ts.tv_nsec = 1000000000;
+      nanosleep(&ts, NULL);
+    }
+
+    if(FD_ISSET(sockfd, &read_fds)) {
+      fprintf(stderr,
+              "rustls wants us to read_tls. First we need to pull some "
+              "bytes from the socket\n");
+
+      /* Read all bytes until we get EAGAIN. Then loop again to wind up in
+         select awaiting the next bit of data. */
+      for(;;) {
+        result = do_read(conn, rconn);
+        if(result == CRUSTLS_DEMO_AGAIN) {
+          break;
+        }
+        else if(result == CRUSTLS_DEMO_CLOSE_NOTIFY) {
+          goto cleanup;
+        }
+        else if(result != CRUSTLS_DEMO_OK) {
+          goto cleanup;
+        }
+      }
+    }
+    if(FD_ISSET(sockfd, &write_fds)) {
+      fprintf(stderr, "rustls wants us to write_tls.\n");
+      err = rustls_connection_write_tls(rconn, write_cb, conn, &n);
+      if(err != 0) {
+        fprintf(stderr, "Error in write_tls: errno %d\n", err);
+        goto cleanup;
+      }
+      else if(n == 0) {
+        fprintf(stderr, "EOF from write_tls\n");
+        goto cleanup;
+      }
+    }
+
+    if (state == READING_REQUEST && body_begin(&conn->data) != NULL) {
+      state = SENT_RESPONSE;
+      fprintf(stderr, "writing response\n");
+      send_response(conn);
+    }
+  }
+
+  fprintf(stderr, "handle_conn: loop fell through");
+
+cleanup:
+  if(sockfd > 0) {
+    close(sockfd);
+  }
+  return NULL;
+}
+
+const struct rustls_certified_key *
+load_cert_and_key(const char* certfile, const char *keyfile) {
+  char certbuf[10000];
+  size_t certbuf_len;
+  char keybuf[10000];
+  size_t keybuf_len;
+
+  int result = read_file(certfile, certbuf, sizeof(certbuf), &certbuf_len);
+  if(result != CRUSTLS_DEMO_OK) {
+    return NULL;
+  }
+
+  result = read_file(keyfile, keybuf, sizeof(keybuf), &keybuf_len);
+  if(result != CRUSTLS_DEMO_OK) {
+    return NULL;
+  }
+
+  const struct rustls_certified_key *certified_key;
+  result = rustls_certified_key_build(
+    (uint8_t*)certbuf, certbuf_len, (uint8_t*)keybuf, keybuf_len, &certified_key);
+  if(result != RUSTLS_RESULT_OK) {
+    print_error("parsing certificate and key", result);
+    return NULL;
+  }
+  return certified_key;
+}
+
+int
+main(int argc, const char **argv)
+{
+  int ret = 1;
+  int result = 1;
+  const char *certfile = argv[1];
+  const char *keyfile = argv[2];
```

Comment:
It's not safe to access elements of `argv` before you check `if(argc <=2)`, below.